rs-matter-macros-impl/src/tlv.rs

```diff
@@ -0,0 +1,511 @@
+/// structure, sequentially, will get Context tags starting from 0
+/// Some configurations are possible through the 'tlvargs' attributes.
+/// For example:
+///  #[tlvargs(start = 1, datatype = "list")]
+///
+/// start: This can be used to override the default tag from which the
+///        encoding starts (Default: 0)
+/// datatype: This can be used to define whether this data structure is
+///        to be encoded as a structure or list. Possible values: list
+///        (Default: struct)
+///
+/// Additionally, structure members can use the tagval attribute to
+/// define a specific tag to be used
+/// For example:
+///  #[tagval(22)]
+///  name: u8,
+/// In the above case, the 'name' attribute will be encoded/decoded with
+/// the tag 22
+pub fn derive_totlv(ast: DeriveInput, crate_name: String) -> TokenStream {
+    let name = &ast.ident;
+
+    let tlvargs = parse_tlvargs(&ast, crate_name);
+    let generics = ast.generics;
+
+    if let syn::Data::Struct(syn::DataStruct {
+        fields: syn::Fields::Named(ref fields),
+        ..
+    }) = ast.data
+    {
+        gen_totlv_for_struct(fields, name, &tlvargs, &generics)
+    } else if let syn::Data::Enum(data_enum) = ast.data {
+        gen_totlv_for_enum(&data_enum, name, &tlvargs, &generics)
+    } else {
+        panic!(
+            "Derive ToTLV - Only supported Struct for now {:?}",
+            ast.data
+        );
+    }
+}
+
+/// Generate a FromTlv implementation for a structure
+fn gen_fromtlv_for_struct(
+    fields: &syn::FieldsNamed,
+    struct_name: &proc_macro2::Ident,
+    tlvargs: TlvArgs,
+    generics: &syn::Generics,
+) -> TokenStream {
+    let mut tag_start = tlvargs.start;
+    let lifetime = tlvargs.lifetime;
+    let datatype = format_ident!("confirm_{}", tlvargs.datatype);
+
+    let mut idents = Vec::new();
+    let mut types = Vec::new();
+    let mut tags = Vec::new();
+
+    for field in fields.named.iter() {
+        let type_name = &field.ty;
+        if let Some(a) = parse_tag_val(field) {
+            // TODO: The current limitation with this is that a hard-coded integer
+            // value has to be mentioned in the tagval attribute. This is because
+            // our tags vector is for integers, and pushing an 'identifier' on it
+            // wouldn't work.
+            tags.push(a);
+        } else {
+            tags.push(tag_start);
+            tag_start += 1;
+        }
+        idents.push(&field.ident);
+
+        if let Type::Path(path) = type_name {
+            types.push(&path.path.segments[0].ident);
+        } else {
+            panic!("Don't know what to do {:?}", type_name);
+        }
+    }
+
+    let krate = Ident::new(&tlvargs.crate_name, Span::call_site());
+
+    // Currently we don't use find_tag() because the tags come in sequential
+    // order. If ever the tags start coming out of order, we can use find_tag()
+    // instead
+    if !tlvargs.unordered {
```

Comment:
nit: For the other code generation functions the code is just copied, in the case of this function the `expanded` variable is removed, any particular reason? 
We could either remove for all, or keep for all?

Reply:
I had a clippy change moved code around and that got me to clean it up more than expected.

I will remove them from everywhere: I believe if we want to print things out then `eprintln` should be used and we now have the ability to unit test to validate what we output.

Reply:
Indeed, the tests would be much faster for iteration